src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py

```diff
@@ -577,6 +586,152 @@ def _update_local_map_and_pose(
 
         return current_map, current_pose
 
+    def cluster_binary_map(self, binary_map):
+        labels = measure.label(
+            binary_map.cpu().numpy(), connectivity=2
+        )  # Perform connected component analysis
+        return torch.tensor(labels, dtype=torch.int64, device=binary_map.device)
+
+    def update_instances_in_global_map(
+        self, global_instances, local_map, x_range, y_range
+    ):
+        """Adds new instances from the local map to the global map
+        Arguments:
+        global_map:
+        local_map:
+        xrange:
+        yrange:
+        """
+        p = self.padding_for_instance_overlap  # default: 1
+        d = self.dilation_for_instances  # default: 0
+
+        H = global_instances.shape[0]
+        W = global_instances.shape[1]
+
+        x1, x2 = x_range
+        y1, y2 = y_range
+
+        # padding added on each side
+        t_p = min(x1, p)
+        b_p = min(H - x2, p)
+        l_p = min(y1, p)
+        r_p = min(W - y2, p)
+
+        # the indices of the padded local_map in the global map
+        x_start = x1 - t_p
+        x_end = x2 + b_p
+        y_start = y1 - l_p
+        y_end = y2 + r_p
+
+        local_map = torch.round(local_map)
+
+        # pad the local map
+        extended_local_map = F.pad(local_map.float(), (l_p, r_p), mode="replicate")
+        extended_local_map = F.pad(
+            extended_local_map.transpose(1, 0), (t_p, b_p), mode="replicate"
+        ).transpose(1, 0)
+
+        self.instance_dilation_selem = skimage.morphology.disk(d)
+        # dilate the extended local map
+        if d > 0:
+            extended_dilated_local_map = torch.round(
+                torch.tensor(
+                    cv2.dilate(
+                        extended_local_map.cpu().numpy(),
+                        self.instance_dilation_selem,
+                        iterations=1,
+                    ),
+                    device=local_map.device,
+                    dtype=local_map.dtype,
+                )
+            )
+        else:
+            extended_dilated_local_map = torch.clone(extended_local_map)
+
+        # Determine instances on the extended local map
+        extended_local_labels = self.cluster_binary_map(extended_dilated_local_map)
+
+        # Get the instances from the global map within the local map's region
+        global_instances_within_local = global_instances[x_start:x_end, y_start:y_end]
+
+        # Create a mapping of local instance IDs to global instance IDs
+        instance_mapping = {}
+
+        max_instance_id = torch.max(global_instances)
+
+        # Associate instances in the local map with corresponding instances in the global map
+        for local_instance_id in torch.unique(extended_local_labels):
```

Comment:
This is a very long function that would probably be easier to follow if it were broken up into a couple sub-functions - e.g. one that populates the instance_mapping dict here

Reply:
Separated